wasm_codec/derive/src/lib.rs

```diff
@@ -343,182 +395,195 @@ enum ErrType {
-        abort!(fn_arg, "`self` arguments are forbidden");
-    };
-
-    pat_type
+fn extract_type_from_fn_arg(emitter: &mut Emitter, fn_arg: syn2::FnArg) -> Option<syn2::PatType> {
+    if let syn2::FnArg::Typed(pat_type) = fn_arg {
+        Some(pat_type)
+    } else {
+        emit!(emitter, fn_arg, "`self` arguments are forbidden");
+        None
+    }
 }
 
 fn classify_params_and_state(
-    params: &Punctuated<syn::FnArg, syn::Token![,]>,
-) -> (Option<syn::Type>, Option<syn::Type>) {
+    emitter: &mut Emitter,
+    params: &Punctuated<syn2::FnArg, syn2::Token![,]>,
+) -> Option<(Option<syn2::Type>, Option<syn2::Type>)> {
     match params.len() {
-        0 => (None, None),
+        0 => Some((None, None)),
         1 => {
             let mut params_iter = params.iter();
-            let first_param = extract_type_from_fn_arg(params_iter.next().unwrap().clone());
+            let first_param =
+                extract_type_from_fn_arg(emitter, params_iter.next().unwrap().clone())?;
 
             if let Ok(state_ty) = parse_state_param(&first_param) {
-                (None, Some(state_ty.clone()))
+                Some((None, Some(state_ty.clone())))
             } else {
-                (Some(first_param.ty.deref().clone()), None)
+                Some((Some(first_param.ty.deref().clone()), None))
             }
         }
         2 => {
             let mut params_iter = params.iter();
-            let first_param = extract_type_from_fn_arg(params_iter.next().unwrap().clone());
+            let first_param =
+                extract_type_from_fn_arg(emitter, params_iter.next().unwrap().clone())?;
 
-            let second_param = extract_type_from_fn_arg(params_iter.next().unwrap().clone());
-            match parse_state_param(&second_param) {
-                Ok(state_ty) => (Some(first_param.ty.deref().clone()), Some(state_ty.clone())),
-                Err(diagnostic) => diagnostic.abort(),
-            }
+            let second_param =
+                extract_type_from_fn_arg(emitter, params_iter.next().unwrap().clone())?;
+
+            let Some(state_ty) = emitter.handle(parse_state_param(&second_param)) else {
+                return None;
+            };
+            Some((Some(first_param.ty.deref().clone()), Some(state_ty.clone())))
+        }
+        _ => {
+            emit!(emitter, params, "No more than 2 parameters are allowed");
+            None
         }
-        _ => abort!(params, "No more than 2 parameters are allowed"),
     }
 }
 
-fn parse_state_param(param: &syn::PatType) -> Result<&syn::Type, Diagnostic> {
-    let syn::Pat::Ident(pat_ident) = &*param.pat else {
-        return Err(diagnostic!(
-            param,
-            Level::Error,
-            "State parameter should be an ident"
-        ));
+fn parse_state_param(param: &syn2::PatType) -> Result<&syn2::Type> {
+    let syn2::Pat::Ident(pat_ident) = &*param.pat else {
+        bail!(param, "State parameter should be an ident");
```

Comment:
Maybe I don't understand something here, but why here you use `bail!` with `Result` without any `emitter` and `Option`, but you don't do that in other places.

I'm asking because it looks kind of strange to insert additional `emitter` parameter and `Option` return type

Reply:
The idea of using an emitter and an option is to try and collect as much errors as possible instead of erroring out.

Ideally, one would even try to generate some code with the incomplete information, though it is not always feasible from the complexity standpoint.

For that purpose I think that this function (and other ones that were returning a `Result<..., Diagnostic>`) should be refactored to use the same style.

Reply:
I wanted to use `bail!` only in places where it made sense to stop.
But as @DCNick3 pointed out I'm basically doing the same thing with the `?` operator where I could've gathered more errors.
I'm going to redo / rearrange some checks to do that.